### ISSUE TITLE
network config grains: DNS servers

### ISSUE THREAD

**Issue by notpeter** [saltstack/salt#26860]
It'd be really nice if salt exposed a dns_server grain, maybe with the simple form of:
````yaml
dns_servers:
  - 10.0.0.2
  - 8.8.8.8
````
On unixes this this can be sourced from /etc/resolv.conf, with one line per name server with one of these two variants:
`nameserver 8.8.8.8`
`nameserver 2001:4860:4860::8844` (still quite rare, even in ipv6 deployments)

On windows the [FIXED_INFO](https://msdn.microsoft.com/en-us/library/aa365900(v=vs.85).aspx) struct returned from [GetNetworkParams](https://msdn.microsoft.com/en-us/library/aa365968(VS.85).aspx) includes the list of currently configured DNS servers.

**Comment by GreatSnoopy** [comment]
You can write your own grain to do that. Start for example, from :
```python

#!/usr/bin/env python

'''
    Very basic grain to return dns configuration.
    It only works in *nix
    It does not yet work with "new style" desktop configurations that use a local instance of
    dnsmasq (it will return 127.0.0.1 as the nameserver)
'''

from __future__ import absolute_import
import re
import salt.utils




def dns_configuration():
    grains={}
    if salt.utils.is_windows():
        return grains
    else:
        try:
            with salt.utils.fopen('/etc/resolv.conf', 'r') as _fh:
                relevant_lines=[ l.strip() for l in _fh.readlines() if not  re.match('^\s*#.*$',l) ]
                dns_servers=[ matches.group(1)  for matches in  [ re.match('^\s*nameserver\s+([.\w]+).*',s) for s in relevant_lines ] if matches ]
                search_domains_lines=[l for l in relevant_lines if re.match('^\s*search\s+\w+.*',l)]
                if len(search_domains_lines) > 0:
                    search_domains=re.split('\s+',search_domains_lines[-1:][0])[1:]
                else:
                    search_domains=[]
        except IOError:
            return grains
    grains['dns_servers']=dns_servers
    grains['dns_search_domains']=search_domains
    return grains
```

**Comment by bersace** [comment]
:+1: for resolv.conf in grains :-)

**Comment by jtand** [comment]
@notpeter Thanks for the feature request!

**Comment by silenius** [comment]
This ticket can be closed, Grains DNS have been added in #27621 :)

**Comment by notpeter** [comment]
Thanks all, especially @silenius! Looking at the code, the dns grain in 2016.3 will be a dict with the following keys:

dns:
  nameservers
  ip4_nameservers
  ip6_nameservers
  domain
  search

Perfect!
